pallets/utxo/src/script.rs

```diff
@@ -95,56 +58,57 @@ impl chainscript::Context for MLContext<'_> {
     ///   The pubkey with unknown type ID always succeeds validation. This allows the type to be
     ///   allocated later for a new signature scheme without introducing a hard fork.
     /// * `Some(Ok(pk))` is a successfully processed key.
-    fn parse_pubkey(&self, pk: &[u8]) -> Option<Self::Public> {
-        let &key_type = pk.get(0)?;
-        if (key_type as usize) < Public::VARIANT_COUNT {
-            Public::decode_all(&mut &pk[..]).map(Ok).ok()
-        } else {
-            Some(Err(key_type))
-        }
+    fn parse_pubkey(&self, pk: &[u8]) -> ParseResult<Self::Public> {
+        Self::Public::parse(pk)
+    }
+
+    /// Extract a signature and sighash.
+    fn parse_signature(&self, pk: Self::Public, sig: &[u8]) -> Option<Self::SignatureData> {
+        pk.parse_sig(sig)
     }
 
     /// Verify signature.
-    fn verify_signature(
-        &self,
-        sig: &Self::Signature,
-        pk: &Self::Public,
-        _subscript: &[u8],
-    ) -> bool {
-        let Signature {
-            signature: sig,
-            sighash: _sighash,
-        } = sig;
-        match pk {
-            Ok(Public::Schnorr(pk)) => (&sig[..]).try_into().map_or(false, |sig| {
-                crypto::sr25519_verify(&sr25519::Signature::from_raw(sig), self.tx_data, pk)
-            }),
-            // Unrecognized signature type => accept the signature.
-            Err(_pk) => true,
-        }
+    fn verify_signature(&self, sd: &Self::SignatureData, _: &[u8], sep_idx: u32) -> bool {
```

Comment:
What is the third parameter and why is it not used?

Reply:
It has to do with `OP_CODESEPARATOR`. Bitcoin uses it in two ways, depending on script version.
1. The portion of the script starting at the last seen CODESEPATAOR is reinjected into the transaction input to be signed. This is before taproot.
2. The position of the last seen `OP_CODESEPARATOR` in message to be signed. This is used in Taproot script and is way simpler and more efficient.

The missing parameter gives you the script fragment from the last code separator onwards. We only implement the second approach so this is not needed. However, I like to keep the `chainscript` crate generic enough for it to be possible to implement the Bitcoin semantics on top of that too.

In general `OP_CODESEPARATOR` is not used very much, I did not find any decent examples with it, and value it brings I reckon is a bit questionable.